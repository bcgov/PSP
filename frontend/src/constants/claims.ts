/**
 * Claims enum, provides a list of permissions that govern what actions are available to an authenticated user.
 */
export enum Claims {
  CONTACT_EDIT = 'contact-edit',
  CONTACT_VIEW = 'contact-view',
  CONTACT_CREATE = 'contact-create',
  PROPERTY_VIEW = 'property-view',
  PROPERTY_EDIT = 'property-edit',
  PROPERTY_ADD = 'property-add',
  PROPERTY_DELETE = 'property-delete',
  ADMIN_USERS = 'admin-users',
  ADMIN_ROLES = 'admin-roles',
  ADMIN_PROPERTIES = 'admin-properties',
  ADMIN_PROJECTS = 'admin-projects',
}

export default Claims;
