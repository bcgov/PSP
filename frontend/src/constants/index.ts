export * from './accessStatus';
export * from './addressTypes';
export * from './claims';
export * from './environment';
export * from './leaseStatusTypes';
export * from './organizationIdentifierTypes';
export * from './organizationTypes';
export * from './projectRiskTypes';
export * from './projectStatusTypes';
export * from './projectTierTypes';
export * from './projectTypes';
export * from './propertyAreaUnitTypes';
export * from './propertyClassificationTypes';
export * from './propertyDataSourceTypes';
export * from './propertyServiceFileTypes';
export * from './propertyStatusTypes';
export * from './propertyTenureTypes';
export * from './propertyTypes';
export * from './roles';
