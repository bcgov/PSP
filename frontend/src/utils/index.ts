export * from './getCurrentFiscal';
export * from './getCurrentYearEvaluation';
export * from './getMostRecentAppraisal';
export * from './getMostRecentEvaluation';
export * from './mapLookupCode';
export * from './mockKeycloak';
export * from './numberFormatUtils';
export * from './propertyUtils';
export * from './utils';
