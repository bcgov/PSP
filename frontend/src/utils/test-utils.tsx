import { useKeycloak } from '@react-keycloak/web';
import {
  fireEvent,
  render as rtlRender,
  RenderOptions as RtlRenderOptions,
  RenderResult,
} from '@testing-library/react';
import { FilterProvider } from 'components/maps/providers/FIlterProvider';
import { createMemoryHistory, MemoryHistory } from 'history';
import noop from 'lodash/noop';
import React, { ReactNode } from 'react';
import { MapContainer } from 'react-leaflet';
import { Router } from 'react-router-dom';
import { ToastContainer } from 'react-toastify';

import TestCommonWrapper from './TestCommonWrapper';

export const mockKeycloak = (
  props: {
    claims?: string[];
    roles?: string[];
    organizations?: number[];
    authenticated?: boolean;
  } = {},
) => {
  const { claims, roles, organizations, authenticated = true } = props;
  // mock what would be returned by keycloak userinfo endpoint
  const userInfo = {
    organizations: organizations ?? [1],
    roles: claims ?? [],
    groups: roles ?? [],
    email: 'test@test.com',
    name: 'Chester Tester',
  };

  (useKeycloak as jest.Mock).mockReturnValue({
    keycloak: {
      userInfo,
      subject: 'test',
      authenticated,
      loadUserInfo: jest.fn().mockResolvedValue(userInfo),
    },
  });
};

export const fillInput = async (
  container: HTMLElement,
  name: string,
  value: any,
  type: string = 'input',
) => {
  let input: Element | null = null;

  if (type === 'radio') {
    input = container.querySelector(`#input-${name}`);
  } else {
    if (type === 'typeahead' || type === 'datepicker') {
      input = container.querySelector(`input[name="${name}"]`);
    } else {
      input = container.querySelector(`${type}[name="${name}"]`);
    }
  }

  // abort early if no input field found
  if (!input) return { input };
  if (type === 'input') {
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
    fireEvent.focusOut(input);
  } else if (type === 'typeahead') {
    fireEvent.focus(input);
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
    const select = container.querySelector(`[aria-label="${value}"]`);
    fireEvent.click(select!);
    fireEvent.focusOut(input);
  } else if (type === 'datepicker') {
    fireEvent.mouseDown(input);
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
    fireEvent.keyPress(input, { key: 'Enter', code: 'Enter' });
  } else if (type === 'radio') {
    fireEvent.click(input);
    fireEvent.focusOut(input);
  } else {
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
    fireEvent.focusOut(input);
  }
  fireEvent.blur(input);

  return { input };
};

export const flushPromises = () => new Promise(setImmediate);

export const deferred = () => {
  let resolve: (value?: unknown) => void = noop;
  const promise = new Promise(_resolve => {
    resolve = _resolve;
  });
  return {
    resolve,
    promise,
  };
};

/**
 * Utility type for generic props of component testing
 */
export interface PropsWithChildren {
  children?: ReactNode;
}

const defaultHistory = createMemoryHistory({
  getUserConfirmation: (message, callback) => callback(true),
});

/**
 * Creates a in-memory router for unit testing
 * @param history (optional) a memory history to use instead of default
 * @returns The route provider
 */
export function createRouteProvider(history?: MemoryHistory) {
  return function Wrapper({ children }: PropsWithChildren) {
    return <Router history={history ?? defaultHistory}>{children}</Router>;
  };
}

/**
 * Creates a Map wrapper for unit testing
 * @param done A callback that will be called when the map has finished rendering
 * @returns The map container instance
 *
 * @example
 *    const { promise, resolve } = deferred();
 *    render(<TestComponent/>, { wrapper: createMapContainer(resolve) })
 *    await waitFor(() => promise)
 *    // the map is fully initialized here...
 */
export function createMapContainer(
  done: () => void = noop,
  whenCreated: (map: L.Map) => void = noop,
) {
  return function Container({ children }: PropsWithChildren) {
    return (
      <div id="mapid" style={{ width: 500, height: 500 }}>
        <MapContainer
          center={[48.43, -123.37]}
          zoom={14}
          whenReady={done}
          whenCreated={whenCreated}
        >
          {children}
        </MapContainer>
      </div>
    );
  };
}

// re-export everything from RTL
export * from '@testing-library/react';

// override RenderOptions interface
export interface RenderOptions extends RtlRenderOptions {
  store?: any;
  history?: MemoryHistory;
  useMockAuthentication?: boolean;
  organizations?: number[];
  claims?: string[];
  roles?: string[];
}

function render(
  ui: React.ReactElement,
  options: Omit<RenderOptions, 'wrapper'> = {},
): RenderResult {
  const {
    store,
    history,
    useMockAuthentication = false,
    organizations,
    claims,
    roles,
    ...renderOptions
  } = options;

  // mock authentication state prior to rendering. Check first that keycloak has been mocked!
  if (!!useMockAuthentication || !!claims || !!roles || !!organizations) {
    if (typeof (useKeycloak as jest.Mock).mockReturnValue === 'function') {
      mockKeycloak({
        claims: claims ?? [],
        roles: roles ?? [],
        organizations: organizations ?? [1],
        authenticated: true,
      });
    }
  }

  function AllTheProviders({ children }: PropsWithChildren) {
    return (
      <TestCommonWrapper store={store} history={history}>
        <ToastContainer
          autoClose={5000}
          hideProgressBar
          newestOnTop={false}
          closeOnClick={false}
          rtl={false}
          pauseOnFocusLoss={false}
        />
        <FilterProvider>{children}</FilterProvider>
      </TestCommonWrapper>
    );
  }
  return rtlRender(ui, { wrapper: AllTheProviders, ...renderOptions });
}

async function renderAsync(
  ui: React.ReactElement,
  options: Omit<RenderOptions, 'wrapper'> = {},
): Promise<RenderResult> {
  const {
    store,
    history,
    useMockAuthentication = false,
    organizations,
    claims,
    roles,
    ...renderOptions
  } = options;

  // mock authentication state prior to rendering. Check first that keycloak has been mocked!
  if (!!useMockAuthentication || !!claims || !!roles || !!organizations) {
    if (typeof (useKeycloak as jest.Mock).mockReturnValue === 'function') {
      mockKeycloak({
        claims: claims ?? [],
        roles: roles ?? [],
        organizations: organizations ?? [1],
        authenticated: true,
      });
    }
  }

  function AllTheProviders({ children }: PropsWithChildren) {
    return (
      <TestCommonWrapper store={store} history={history}>
        <ToastContainer
          autoClose={5000}
          hideProgressBar
          newestOnTop={false}
          closeOnClick={false}
          rtl={false}
          pauseOnFocusLoss={false}
        />
        <FilterProvider>{children}</FilterProvider>
      </TestCommonWrapper>
    );
  }
  return await rtlRender(ui, { wrapper: AllTheProviders, ...renderOptions });
}

// override render method
export { render, renderAsync };
