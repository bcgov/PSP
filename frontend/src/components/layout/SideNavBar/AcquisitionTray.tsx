import { Claims } from 'constants/claims';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import { Link } from 'react-router-dom';

import { ISideTrayPageProps } from './SideTray';
import * as Styled from './styles';

/**
 * Acquisitions page with links to acquisition functionality.
 * Intended for use within the left side tray.
 */
export const AcquisitionTray = ({ onLinkClick }: ISideTrayPageProps) => {
  const { hasClaim } = useKeycloakWrapper();
  return (
    <>
      <Styled.TrayHeader>Acquisition Files</Styled.TrayHeader>
      {hasClaim(Claims.ACQUISITION_VIEW) && (
        <Link onClick={onLinkClick} to="/acquisition/list">
          Manage Acquisition files
        </Link>
      )}
      {hasClaim(Claims.ACQUISITION_ADD) && (
        <Link onClick={onLinkClick} to="/mapview/sidebar/acquisition/new">
          Create an Acquisition file
        </Link>
      )}
    </>
  );
};
