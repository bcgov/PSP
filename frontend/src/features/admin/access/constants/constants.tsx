import { ColumnWithProps } from 'components/Table';
import { AccessRequestStatus, AccessStatusDisplayMapper } from 'constants/accessStatus';
import { AccessRequestForm } from 'features/admin/access-request/models';
import * as React from 'react';
import { Link } from 'react-router-dom';
import { CellProps } from 'react-table';

import { RowActions } from '../components/RowActions';

export const getAccessRequestColumns = (
  refresh: () => void,
): ColumnWithProps<AccessRequestForm>[] => [
  {
    Header: 'IDIR/BCeID',
    accessor: 'businessIdentifierValue',
    align: 'left',
    Cell: (props: CellProps<AccessRequestForm>) => {
      return (
        <Link to={`/admin/access/request/${props.row.original.id}`}>
          {props.row.original.businessIdentifierValue}
        </Link>
      );
    },
  },
  {
    Header: 'First name',
    accessor: 'firstName',
    align: 'left',
  },
  {
    Header: 'Last name',
    accessor: 'surname',
    align: 'left',
  },
  {
    Header: 'Email',
    accessor: 'email',
    align: 'left',
    minWidth: 200,
    maxWidth: 300,
  },
  {
    Header: 'Position',
    accessor: 'position',
    align: 'left',
  },
  {
    Header: 'Status',
    accessor: 'accessRequestStatusTypeCodeId',
    align: 'left',
    width: 100,
    Cell: (props: CellProps<AccessRequestForm>) => {
      return AccessStatusDisplayMapper[
        props.row.original.accessRequestStatusTypeCodeId as AccessRequestStatus
      ];
    },
  },
  {
    Header: 'Role',
    accessor: 'roleName',
    align: 'left',
    minWidth: 200,
  },
  {
    Header: 'MoTI Region',
    accessor: 'regionName',
    align: 'left',
    minWidth: 200,
  },
  {
    Header: 'Actions',
    Cell: (props: CellProps<AccessRequestForm>) => {
      return <RowActions {...props} refresh={refresh} />;
    },
    width: 75,
  },
];
