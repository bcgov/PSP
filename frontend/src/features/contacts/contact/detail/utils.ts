import { IContactAddress } from 'interfaces/IContact';

import { AddressTypes } from './../../../../constants/addressTypes';

// the order of this array corresponds to the expected display order
const addressSortOrder = [AddressTypes.Mailing, AddressTypes.Residential, AddressTypes.Billing];

export const sortAddresses = (a1: IContactAddress, a2: IContactAddress) => {
  if (a2.addressType.id === a1.addressType.id) {
    return 0;
  }
  const a2Index = addressSortOrder.indexOf(a2.addressType.id as AddressTypes);
  if (a2Index !== -1 && a2Index < addressSortOrder.indexOf(a1.addressType.id as AddressTypes)) {
    return 1;
  }

  return -1;
};

export const fakeAddresses = [
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'BILLING',
      description: 'Billing address',
      isDisabled: false,
    },
    streetAddress1: 'Billing address',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6A 5G7',
  },
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'MAILING',
      description: 'Mailing address',
      isDisabled: false,
    },
    streetAddress1: 'Mailing address',
    streetAddress2: 'Living in a van',
    streetAddress3: 'Down by the River',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6Z 5G7',
  },
  {
    id: 5,
    rowVersion: 1,
    addressType: {
      id: 'RESIDNT',
      description: 'Property address',
      isDisabled: false,
    },
    streetAddress1: 'Property address',
    streetAddress2: '',
    streetAddress3: '',
    municipality: 'vic',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V9A 0H6',
  },
];
