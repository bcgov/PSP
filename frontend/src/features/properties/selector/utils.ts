import { IMapProperty } from 'features/properties/selector/models';
import { compact } from 'lodash';
import { Api_PropertyFile } from 'models/api/PropertyFile';
import { pidFormatter } from 'utils';

export enum NameSourceType {
  PID = 'PID',
  PIN = 'PIN',
  PLAN = 'Plan #',
  LOCATION = 'Location',
  NAME = 'Descriptive Name',
  NONE = 'n/a',
}

interface PropertyName {
  label: NameSourceType;
  value: string;
}

export const getPropertyName = (property: IMapProperty): PropertyName => {
  if (property.pid !== undefined && property.pid?.toString().length > 0 && property.pid !== '0') {
    return { label: NameSourceType.PID, value: pidFormatter(property.pid.toString()) };
  } else if (
    property.pin !== undefined &&
    property.pin?.toString()?.length > 0 &&
    property.pin !== '0'
  ) {
    return { label: NameSourceType.PIN, value: property.pin.toString() };
  } else if (property.planNumber !== undefined) {
    return { label: NameSourceType.PLAN, value: property.planNumber };
  } else if (property.latitude !== undefined && property.longitude !== undefined) {
    return {
      label: NameSourceType.LOCATION,
      value: compact([property.latitude?.toFixed(5), property.longitude?.toFixed(5)]).join(', '),
    };
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getFilePropertyName = (
  fileProperty?: Api_PropertyFile,
  skipName: boolean = false,
): PropertyName => {
  if (fileProperty === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  if (
    fileProperty.propertyName !== undefined &&
    fileProperty.propertyName !== '' &&
    skipName === false
  ) {
    return { label: NameSourceType.NAME, value: fileProperty.propertyName };
  } else if (fileProperty.property !== undefined) {
    const property = fileProperty.property;
    let mapProperty: IMapProperty = {
      pin: property.pin?.toString(),
      pid: property.pid?.toString(),
      latitude: property.latitude,
      longitude: property.longitude,
      planNumber: property.planNumber,
    };
    return getPropertyName(mapProperty);
  }
  return { label: NameSourceType.NONE, value: '' };
};
