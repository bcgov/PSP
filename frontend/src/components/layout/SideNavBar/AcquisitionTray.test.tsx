import { screen } from '@testing-library/react';
import { Claims } from 'constants/claims';
import { noop } from 'lodash';
import { mockKeycloak, render, RenderOptions, waitFor } from 'utils/test-utils';

import { AcquisitionTray } from './AcquisitionTray';

// mock auth library
jest.mock('@react-keycloak/web');

// render component under test
const setup = (renderOptions: RenderOptions = {}) => {
  const utils = render(<AcquisitionTray onLinkClick={noop} />, {
    ...renderOptions,
  });
  return { ...utils };
};

describe('AcquisitionTray', () => {
  it('matches snapshot', async () => {
    mockKeycloak({ claims: [Claims.ACQUISITION_VIEW] });
    const { asFragment } = setup();

    const fragment = await waitFor(() => asFragment());
    expect(fragment).toMatchSnapshot();
  });

  it('should have the Acquisition Files header in the component', async () => {
    setup({});
    expect(screen.getByText(`Acquisition Files`)).toBeInTheDocument();
  });

  it(`should have the "Manage acquisition files" link in the component`, async () => {
    mockKeycloak({ claims: [Claims.ACQUISITION_VIEW] });
    setup({});
    expect(screen.getByText(`Manage acquisition files`)).toBeInTheDocument();
  });

  it(`should have the Create an acquisition file link in the component`, async () => {
    mockKeycloak({ claims: [Claims.ACQUISITION_ADD] });
    setup({});
    expect(screen.getByText(`Create an acquisition file`)).toBeInTheDocument();
  });

  it(`should not have the "Manage acquisition files" link in the component`, async () => {
    mockKeycloak({ claims: [] });
    setup({});
    expect(screen.queryByText(`Manage acquisition files`)).toBe(null);
  });

  it(`should not have the "Create an acquisition file" link in the component`, async () => {
    mockKeycloak({ claims: [] });
    setup({});
    expect(screen.queryByText(`Create an acquisition file`)).toBe(null);
  });
});
