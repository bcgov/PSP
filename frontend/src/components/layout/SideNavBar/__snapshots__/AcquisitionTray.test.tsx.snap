// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AcquisitionTray matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="container"
  />
  <div
    class="Toastify"
  />
  .c0 {
  font-size: 3rem;
  padding-bottom: 0.8rem;
  border-bottom: 0.4rem solid;
  margin-bottom: 3.2rem;
  max-width: 28rem;
}

<h3
    class="c0"
  >
    Acquisition Files
  </h3>
  <a
    href="/acquisition/list"
  >
    Manage Acquisition Files
  </a>
</DocumentFragment>
`;
